### PR TITLE
Add ORTModel support for custom tasks

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

`ORTModelForXXX.model` decided valid inputs and outputs of the model's forward method, thus the creation of inputs in the forward method can be abstract, and also the outputs. This would allow the ORTModels to be more flexible.

e.g. In `ORTTrainer`, the evaluation includes __labels__ as input and __loss__ as output. With the PR, it will enable us to replace bare inference sessions with ORTModels more easily.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/onnxruntime/modeling_ort.py [hunk 1]
```diff
@@ -420,7 +420,8 @@ class ORTModelForQuestionAnswering(ORTModel):
 
     def __init__(self, model=None, config=None, **kwargs):
         super().__init__(model, config, **kwargs)
-        # create {name:idx} dict for model outputs
+        # create {name:idx} dict for model inputs and outputs
+        self.model_inputs = {output_key.name: idx for idx, output_key in enumerate(self.model.get_inputs())}
```
Comment (regisss) [inline optimum/onnxruntime/modeling_ort.py:424]:
```suggestion
        self.model_inputs = {input_key.name: idx for idx, input_key in enumerate(self.model.get_inputs())}
```
Same

File: optimum/onnxruntime/modeling_ort.py [hunk 2]
```diff
@@ -520,9 +519,9 @@ class ORTModelForSequenceClassification(ORTModel):
 
     def __init__(self, model=None, config=None, **kwargs):
         super().__init__(model, config, **kwargs)
-        # create {name:idx} dict for model outputs
-        self.model_outputs = {output_key.name: idx for idx, output_key in enumerate(self.model.get_outputs())}
+        # create {name:idx} dict for model inputs and outputs
         self.model_inputs = {output_key.name: idx for idx, output_key in enumerate(self.model.get_inputs())}
```
Comment (regisss) [inline optimum/onnxruntime/modeling_ort.py:523]:
```suggestion
        self.model_inputs = {input_key.name: idx for idx, input_key in enumerate(self.model.get_inputs())}
```
Same

File: optimum/onnxruntime/modeling_ort.py [hunk 3]
```diff
@@ -606,7 +603,8 @@ class ORTModelForTokenClassification(ORTModel):
 
     def __init__(self, model=None, config=None, **kwargs):
         super().__init__(model, config, **kwargs)
-        # create {name:idx} dict for model outputs
+        # create {name:idx} dict for model inputs and outputs
+        self.model_inputs = {output_key.name: idx for idx, output_key in enumerate(self.model.get_inputs())}
```
Comment (regisss) [inline optimum/onnxruntime/modeling_ort.py:607]:
```suggestion
        self.model_inputs = {input_key.name: idx for idx, input_key in enumerate(self.model.get_inputs())}
```
Same

File: optimum/onnxruntime/modeling_ort.py [hunk 4]
```diff
@@ -689,8 +686,9 @@ class ORTModelForCausalLM(ORTModel, GenerationMixin):
 
     def __init__(self, model=None, config=None, **kwargs):
         super().__init__(model, config, **kwargs)
-        # create {name:idx} dict for model outputs
+        # create {name:idx} dict for model inputs and outputs
         self.main_input_name = "input_ids"
+        self.model_inputs = {output_key.name: idx for idx, output_key in enumerate(self.model.get_inputs())}
```
Comment (regisss) [inline optimum/onnxruntime/modeling_ort.py:691]:
```suggestion
        self.model_inputs = {input_key.name: idx for idx, input_key in enumerate(self.model.get_inputs())}
```
Same

File: optimum/onnxruntime/modeling_ort.py [hunk 5]
```diff
@@ -807,7 +807,8 @@ class ORTModelForImageClassification(ORTModel):
 
     def __init__(self, model=None, config=None, **kwargs):
         super().__init__(model, config, **kwargs)
-        # create {name:idx} dict for model outputs
+        # create {name:idx} dict for model inputs and outputs
+        self.model_inputs = {output_key.name: idx for idx, output_key in enumerate(self.model.get_inputs())}
```
Comment (regisss) [inline optimum/onnxruntime/modeling_ort.py:811]:
```suggestion
        self.model_inputs = {input_key.name: idx for idx, input_key in enumerate(self.model.get_inputs())}
```

File: optimum/onnxruntime/modeling_ort.py [hunk 6]
```diff
@@ -346,25 +347,24 @@ def __init__(self, model=None, config=None, **kwargs):
...
+        onnx_outputs = self.model.run(None, onnx_inputs)
+        outputs = dict(
+            map(
+                lambda output_name: (
+                    output_name,
+                    torch.from_numpy(onnx_outputs[self.model_outputs[output_name]]).to(self.device),
+                ),
+                self.model_outputs,
+            )
+        )
```
Comment (philschmid) [inline optimum/onnxruntime/modeling_ort.py:365]: Have you measured the performance difference for this? Also is there way to improve the readability of the code? having nested `dict`, `map` etc. is not so nice.

File: optimum/onnxruntime/modeling_ort.py [hunk 7]
```diff
@@ -346,25 +347,24 @@ def __init__(self, model=None, config=None, **kwargs):
...
         # converts pytorch inputs into numpy inputs for onnx
-        onnx_inputs = {
-            "input_ids": input_ids.cpu().detach().numpy(),
-            "attention_mask": attention_mask.cpu().detach().numpy(),
-        }
-        if token_type_ids is not None:
-            onnx_inputs["token_type_ids"] = token_type_ids.cpu().detach().numpy()
+        onnx_inputs = dict(
+            map(lambda input_name: (input_name, kwargs.pop(input_name).cpu().detach().numpy()), self.model_inputs)
+        )
```
Comment (philschmid) [inline optimum/onnxruntime/modeling_ort.py:354]: Similar for here. What are the performance changes by making this change. Also we plan to move to `IOBindinngs`.  
And i am also not a and of removing the `args` from the `forward` pass.

**Review by philschmid — CHANGES_REQUESTED** [review]
I am not sure if we should make this change for enabling something, which the classes aren't designed for. 
You said 
> e.g. In ORTTrainer, the evaluation includes labels as input and loss as output. With the PR, it will enable us to replace bare inference sessions with ORTModels more easily.

This sounds more that we should use `evaluate` and `pipeline` or the ORTModel in the trainer with the post-processing outside of the model.

File: optimum/onnxruntime/modeling_ort.py [hunk 8]
```diff
@@ -335,7 +335,8 @@ class ORTModelForFeatureExtraction(ORTModel):
 
     def __init__(self, model=None, config=None, **kwargs):
         super().__init__(model, config, **kwargs)
-        # create {name:idx} dict for model outputs
+        # create {name:idx} dict for model inputs and outputs
+        self.model_inputs = {output_key.name: idx for idx, output_key in enumerate(self.model.get_inputs())}
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_ort.py:339]: Good point, will do!

File: optimum/onnxruntime/modeling_ort.py (hunk 6 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_ort.py:365]: IMO what users could probably have as inputs for models in transformers, they shall also be able to have them in the `forward` pass of ORTModels. E.g. for BERT model, we could have [various inputs](https://github.com/huggingface/transformers/blob/main/src/transformers/models/bert/modeling_bert.py#L916), but ORTModels only take `input_ids` and `attention_mask` as default and `token_type_ids` as optional. Although this corresponds to the vanilla definition in `FeatureManager`, we shall enable users to take all possible options in transformers as inputs, in case they have exported their model with customized OnnxConfig.
And this is what I did in the `ORTTrainer.evaluate`, I created [wrappers](https://github.com/huggingface/optimum/blob/main/optimum/onnx/configuration.py#L36) to include labels as input and loss as output. Thus the onnx model for evaluation will have labels and loss, and I don't think that could be possible to use pipeline or evaluate or ORTModel to calculate the extra output in such a situation with a a bare onnx model, as we lost information
```
>>> ort_model = ORTModelForSequenceClassification.from_pretrained(model_id="./assets", file_name="distilbert-base-uncased-wo-labels.onnx")
config.json NOT FOUND in HuggingFace Hub
>>> ort_model.config
PretrainedConfig {
  "transformers_version": "4.20.1"
}
```

Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_ort.py:365]: This refactoring does pretty much the same thing as before just in a more abstract way, as the inputs and outputs are defined once we instantiate the ORTModels. 
I do agree that the mapping is not good as one could put no matter what in the `**kwargs` and iterating the invalid inputs could take computation. What about applying a filter for valid keys?

File: optimum/onnxruntime/modeling_ort.py (hunk 7 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_ort.py:354]: I can keep the args! That makes sense!

Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_ort.py:354]: I didn't not consider the IO binding and just adapted our previous code. I open a PR for the replacement with IO binding for inputs and outputs on devices. You think the changes proposed by this PR would impact the replacement by IO binding?

File: optimum/onnxruntime/modeling_ort.py (hunk 6 above)
Comment (philschmid) [reply optimum/onnxruntime/modeling_ort.py:365]: > IMO what users could probably have as inputs for models in transformers, they shall also be able to have them in the forward pass of ORTModels. 

I don't agree with that. The `ORTModelForXXX` are created for inference. To provide an easy abstraction from `transformers` to use onnxruntime for inference but keep the same API. 

But before we decide on any changes can you benchmark before and after. Since a `dict(map())` can cost quite a lot,  especially for generation models where the forward pass ist called multiple times.

Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_ort.py:365]: For benchmarking, with the following snippet:
```python
tokens = tokenizer(["I love burritos and pizza and pasta!"]*10000, return_tensors="pt")
start_time = time.time()
pred = ort_model(**tokens)
runtime = time.time() - start_time
result = {"Inference_runtime": round(runtime, 4)}
print(result)
```
Previous ORTModel takes `{'Inference_runtime': 16.6773}` and the current one takes `{'Inference_runtime': 16.6948}`. 

I am agreed that `dict(map())` is not the best option in terms of computation speed. But I think making the ORTModels able to take various valid inputs is something that we should enable in terms of flexibility. What about putting back arguments and adding some `if` control flow as what is done with `token_type_ids`?

File: optimum/onnxruntime/modeling_ort.py [hunk 9]
```diff
@@ -283,6 +283,22 @@ def _from_transformers(
         # 3. load normal model
         return cls._from_pretrained(save_dir.as_posix(), **kwargs)
 
+    def prepare_onnx_inputs(self, **kwargs):
```
Comment (JingyaHuang) [inline optimum/onnxruntime/modeling_ort.py:286]: @philschmid What about something like that? This one took `{'Inference_runtime': 16.687}`, closer to what we had compared to my previous try.

File: optimum/onnxruntime/modeling_ort.py [hunk 10]
```diff
@@ -833,3 +781,39 @@ def forward(
         return ImageClassifierOutput(
             logits=torch.from_numpy(outputs[self.model_outputs["logits"]]),
         )
+
+
+class ORTModelForCustomTasks(ORTModel):
```
Comment (philschmid) [inline optimum/onnxruntime/modeling_ort.py:786]: i like it!

**Review by philschmid — APPROVED** [review]
can you add a test and then we should be good. Good idea! ✅ And if some use case emerges out of it we can add new task-specific model classes

**Comment by JingyaHuang** [comment]
> I still do not understand the purpose of this change.
> 
> As mentioned before the `ORTModelForXX` were introduced for inference and not for training. The idea was to be able to have API compatible Model classes, which can be used with `pipelines` without the need to re-write the pre - & post-processing. Also, the idea is to add those inference model classes to other packages, e.g `optimum-intel`. Additionally, we are looking into removing the copying and adding `IOBindings` to reduce latency in the future.
> 
> The changes you suggest:
> 
> * slow down the inference
> * add a lot of complex dynamic code -> which we tried to exclude, that's why we have several `ORTModelForXX` classes rather than one
> * Add support for something `ORTTrainier` training specific.
> 
> The question I have is:
> 
> * Is this change needed for the ORTTrainer? how are we currently doing it?
> * What is the benefit for the customer?

Hi @philschmid, sorry for the late reply. I re-drafted the code, indeed it shouldn't be in other task-specific models as it will slow them down. The basic idea behind the PR is to leave some flexibility to users, it's like a fallback so that when they are using a more customized model they can still be able to benefit from the ORTModel foundation with a small sacrifice of speed.
`ORTTrainer.train()` is independent of the PR, but for the inference of `ORTTrainer` the `evaluate()` and `predict()`, I am using directly InferenceSession right now and I will replace it with ORTModel, for the predict it is pretty straight forward, but for the evaluate the model include loss thus I need something more customized, and things like `ORTModelForCustomTasks` shall be helpful.

File: optimum/onnxruntime/modeling_ort.py (hunk 10 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_ort.py:786]: Cool!! I will make it neater, it's a draft idea right now.

Comment (philschmid) [reply optimum/onnxruntime/modeling_ort.py:786]: Awesome! would you mind adding similar documentation to what we have for the others? Maybe you can be creative with a non-supported task/model, but which would work.

Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_ort.py:819]: Example added!

File: optimum/onnxruntime/modeling_ort.py [hunk 11]
```diff
@@ -833,3 +834,86 @@ def forward(
...
+    >>> from transformers import {processor_class}, pipeline
+    >>> from optimum.onnxruntime import {model_class}
+
+    >>> tokenizer = {processor_class}.from_pretrained("{checkpoint}")
+    >>> model = {model_class}.from_pretrained("{checkpoint}")
+    >>> onnx_extractor = pipeline("feature-extraction", model=model, tokenizer=tokenizer)
+
+    >>> text = "I love burritos!"
+    >>> pred = onnx_extractor(text)
+    ```
```
Comment (philschmid) [inline optimum/onnxruntime/modeling_ort.py:869]: Does that make sense? Since have the [ORTModelForFeautreExtraction](https://huggingface.co/docs/optimum/onnxruntime/modeling_ort#optimum.onnxruntime.ORTModelForFeatureExtraction)?

File: optimum/onnxruntime/modeling_ort.py [hunk 12]
```diff
@@ -833,3 +834,86 @@ def forward(
...
+    """
+    Onnx Model for any custom tasks. It can be used to leverage the inference acceleration with any custom exported ONNX model.
+    """,
+    ONNX_MODEL_START_DOCSTRING,
+)
+class ORTModelForCustomTasks(ORTModel):
+    """
+    Onnx Model for any custom tasks.
+    """
+
```
Comment (philschmid) [inline optimum/onnxruntime/modeling_ort.py:883]: Do we need to register it to some Transformers AutoClass similar to the others? that's needed to fix the warnings when pushing into a pipeline

File: optimum/onnxruntime/modeling_ort.py [hunk 13]
```diff
@@ -833,3 +834,86 @@ def forward(
...
+        )
+    )
+    def forward(self, **kwargs):
+        # converts pytorch inputs into numpy inputs for onnx
+        onnx_inputs = self._prepare_onnx_inputs(**kwargs)
+        # run inference
+        onnx_outputs = self.model.run(None, onnx_inputs)
+        outputs = self._prepare_onnx_outputs(onnx_outputs)
+        # converts outputs to namedtuple for pipelines post-processing if applicable
+        return namedtuple("CustomTasksOutput", outputs.keys())(*outputs.values())
```
Comment (philschmid) [inline optimum/onnxruntime/modeling_ort.py:901]: Lets replace this with `ModelOutput` which is the underlying class for all outputs in transformers: Ref: https://github.com/huggingface/transformers/blob/ee67e7ad4fd7a766891b68f708cf03e30f609976/src/transformers/utils/generic.py#L147

**Review by philschmid — APPROVED** [review]
LGTM ✅ Left some minor comments

File: optimum/onnxruntime/modeling_ort.py (hunk 11 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_ort.py:869]: The example here is a [sentence BERT with `last_hidden_state` and `pooler_output`](https://huggingface.co/optimum/sbert-all-MiniLM-L6-with-pooler) as outputs. This can be traced back to [an issue](https://github.com/huggingface/optimum/issues/224) where the user wanted to leverage the `pooler_output` instead of the `last_hidden_state`. The [ORTModelForFeautreExtraction](https://huggingface.co/docs/optimum/onnxruntime/modeling_ort#optimum.onnxruntime.ORTModelForFeatureExtraction) only takes `last_hidden_state` as output so the example with pooler is somehow custom(but maybe not that custom...). WDYT? Otherwise, do you have a better example in mind?

Comment (philschmid) [reply optimum/onnxruntime/modeling_ort.py:869]: Nice!

File: optimum/onnxruntime/modeling_ort.py (hunk 12 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_ort.py:883]: Ok! I will go with 
```python
pipeline_task = "default"
auto_model_class = AutoModel
```

File: optimum/onnxruntime/modeling_ort.py [hunk 14]
```diff
@@ -847,3 +848,88 @@ def forward(
...
         )
+
+
+CUSTOM_TASKS_EXAMPLE = r"""
+    Example of custom tasks(e.g. a sentence transformers taking `pooler_output` as output):
+
+    ```python
+    >>> from transformers import {processor_class}
+    >>> from optimum.onnxruntime import {model_class}
+    >>> import torch
```
Comment (echarlaix) [inline optimum/onnxruntime/modeling_ort.py:859]: torch doesn't seem to be used, could we remove it ?

**Review by echarlaix — APPROVED** [review]
LGTM, great job @JingyaHuang 🔥
